rowexec: add encoded datums to HLL sketch

mgartner opened the pull request:
#### sql: add BenchmarkAnalyze

Release note: None

#### sql/execversion: add V25_2

Release note: None

#### rowexec: add encoded datums to HLL sketch

Previously, the sampler would decode and re-encode all datums except for
integers with EncDatum.Fingerprint before adding them to the hyperloglog
sketch. This ensured that equal bytes were added to the sketch for two
semantically equivalent values, even if their encodings were different.
The overhead of decoding and re-encoding is significant, however.

Now, for all types except for collated strings or container types with
collated strings, the encoded bytes of the datum, if present, are added
directly to the hyperloglog sketch, skipping the overhead of
fingerprinting. See the comments added to `addRow` for more details.

```
name     old time/op    new time/op    delta
Analyze     216ms ± 3%     158ms ± 3%  -27.14%  (p=0.000 n=9+10)

name     old alloc/op   new alloc/op   delta
Analyze    75.6MB ± 4%    63.7MB ± 3%  -15.66%  (p=0.000 n=9+9)

name     old allocs/op  new allocs/op  delta
Analyze      573k ± 5%      355k ± 1%  -37.94%  (p=0.000 n=10+8)
```

Fixes #137362
Fixes #140652

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/140601)
<!-- Reviewable:end -->

mgartner commented:
I measured a slight speed up in `BenchmarkSysbench` with table stats enabled, though hard to say if it's just noise. 

```
name                                  old time/op    new time/op    delta
Sysbench/SQL/3node/oltp_read_write-8    4.96ms ± 0%    4.93ms ± 0%  -0.52%  (p=0.000 n=10+10)

name                                  old alloc/op   new alloc/op   delta
Sysbench/SQL/3node/oltp_read_write-8    2.11MB ± 0%    2.12MB ± 0%  +0.48%  (p=0.000 n=10+9)

name                                  old allocs/op  new allocs/op  delta
Sysbench/SQL/3node/oltp_read_write-8     9.90k ± 0%     9.91k ± 0%    ~     (p=0.225 n=10+10)
```

yuzefovich reviewed (commented):
Nice! The change should be an improvement on a full scale benchmark, but I'd expect this to be noise on a microbenchmark.

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-:-OISwQFu3mhecmDPetOn:behz0v8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @michae2, and @mw5h)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OISKhdqBVtnUo6WWOnY:-OISKhdqBVtnUo6WWOnZ:bb3plce):*
I think we need to do a migration similar to what we did in 62765d6e19b4d1c612a2feeb30561435991a1f36. In particular, if we have a mixed-version cluster, with the current PR version, the sample aggregator could receive marshalled sketches with different contents (old nodes would send fingerprinted datums, new nodes would send encoded).
___
*[`-- commits` line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OISKiER-hqcczZTEJj3:-OISKiER-hqcczZTEJj4:bxhn56z):*
What is special about collated strings? It'd be worth including it into the commit message and probably the comment in the code too.
___
*[`pkg/sql/rowexec/sampler.go` line 520 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OISJHgpDzSymGxSpLyU:-OISJHgpDzSymGxSpLyV:b-ipoakw) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c085b098b605e35121fa361509ec3c7d98add3b/pkg/sql/rowexec/sampler.go#L520)):*
> ```Go
> 			return true
> 		}
> 		if t.Family() == types.ArrayFamily {
> ```

Should we also check for tuples with collated strings? In theory, it should be possible - although in practice it seems not - to create a composite type with a collated string (unsupported due to #132589) or use the implicit table record type (which also doesn't appear to work), so it might be good to be defensive here.
___
*[`pkg/sql/rowexec/sampler_test.go` line 388 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIT-j3yAz3wuig5Nz8B:-OIT-j3yAz3wuig5Nz8C:bdt25ka) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c085b098b605e35121fa361509ec3c7d98add3b/pkg/sql/rowexec/sampler_test.go#L388)):*
> ```Go
> 						outTypes[0] = tc.typs[1]
> 					case types.FloatFamily:
> 						inputLen = len(inputRows)
> ```

nit: do we not need to override `outTypes` here to float? I would expect that we don't need it only for `[][]int` case because the original `outTypes` contents matches then. I'd also consider keeping `outTypes[0]` and `outTypes[1]` be left as nil where `outTypes` is defined so that each case had to explicitly set the types.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Very nice!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-:-OIXWSqCAiFVTPQcdt6a:b-j7qk9t)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @mw5h)

___
*[`pkg/sql/rowexec/sampler.go` line 518 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIXTqYxCuMZj91sc_1t:-OIXTqYxCuMZj91sc_1u:b-ijzrfg) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c085b098b605e35121fa361509ec3c7d98add3b/pkg/sql/rowexec/sampler.go#L518)):*
> ```Go
> 	containsCollatedString := func(t *types.T) bool {
> 		if t.Family() == types.CollatedStringFamily {
> 			return true
> ```

nit: might also be worth checking if the collation differs from `default` / `IsDefaultEquivalentCollation()`


<!-- Sent from Reviewable.io -->

mgartner committed:
sql: add BenchmarkAnalyze

Release note: None

mgartner committed:
sql/execversion: add V25_2

Release note: None

mgartner committed:
rowexec: add encoded datums to HLL sketch

Previously, the sampler would decode and re-encode all datums except for
integers with EncDatum.Fingerprint before adding them to the hyperloglog
sketch. This ensured that equal bytes were added to the sketch for two
semantically equivalent values, even if their encodings were different.
The overhead of decoding and re-encoding is significant, however.

Now, for all types except for collated strings or container types with
collated strings, the encoded bytes of the datum, if present, are added
directly to the hyperloglog sketch, skipping the overhead of
fingerprinting. See the comments added to `addRow` for more details.

```
name     old time/op    new time/op    delta
Analyze     216ms ± 3%     158ms ± 3%  -27.14%  (p=0.000 n=9+10)

name     old alloc/op   new alloc/op   delta
Analyze    75.6MB ± 4%    63.7MB ± 3%  -15.66%  (p=0.000 n=9+9)

name     old allocs/op  new allocs/op  delta
Analyze      573k ± 5%      355k ± 1%  -37.94%  (p=0.000 n=10+8)
```

Fixes #137362
Fixes #140652

Release note: None

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140601)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mw5h and @yuzefovich)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OISKhdqBVtnUo6WWOnY:-OIpOtMx9TvayMr2fhAP:b-896fix):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I think we need to do a migration similar to what we did in 62765d6e19b4d1c612a2feeb30561435991a1f36. In particular, if we have a mixed-version cluster, with the current PR version, the sample aggregator could receive marshalled sketches with different contents (old nodes would send fingerprinted datums, new nodes would send encoded).
</blockquote></details>

Done.
___
*[`-- commits` line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OISKiER-hqcczZTEJj3:-OIq-xbKAgTfygvkSJT1:b-896fix):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

What is special about collated strings? It'd be worth including it into the commit message and probably the comment in the code too.
</blockquote></details>

Done.
___
*[`pkg/sql/rowexec/sampler.go` line 518 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIXTqYxCuMZj91sc_1t:-OIpEW5E8vFZatLpSo-r:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c085b098b605e35121fa361509ec3c7d98add3b/pkg/sql/rowexec/sampler.go#L518)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

nit: might also be worth checking if the collation differs from `default` / `IsDefaultEquivalentCollation()`
</blockquote></details>

Done.
___
*[`pkg/sql/rowexec/sampler.go` line 520 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OISJHgpDzSymGxSpLyU:-OIpEWqh673M_XAeczU1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c085b098b605e35121fa361509ec3c7d98add3b/pkg/sql/rowexec/sampler.go#L520)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Should we also check for tuples with collated strings? In theory, it should be possible - although in practice it seems not - to create a composite type with a collated string (unsupported due to #132589) or use the implicit table record type (which also doesn't appear to work), so it might be good to be defensive here.
</blockquote></details>

Done.
___
*[`pkg/sql/rowexec/sampler_test.go` line 388 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIT-j3yAz3wuig5Nz8B:-OIpF-j0CdONKwd-Pl53:bfywx0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c085b098b605e35121fa361509ec3c7d98add3b/pkg/sql/rowexec/sampler_test.go#L388)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: do we not need to override `outTypes` here to float? I would expect that we don't need it only for `[][]int` case because the original `outTypes` contents matches then. I'd also consider keeping `outTypes[0]` and `outTypes[1]` be left as nil where `outTypes` is defined so that each case had to explicitly set the types.
</blockquote></details>

The output types are never actually used in this test. We only need the length of `outTypes` to be correct, so I've simplified things and explained that.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
Nice improvement! Just a handful of nits from me :smiley: <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 8 of 8 files at r2, 5 of 5 files at r3, 7 of 7 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-:-OIrrDEV175eJlUVGGIN:b-zhmahs)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mw5h)

___
*[`pkg/sql/create_stats_test.go` line 226 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIrnIPmBtQOmu0XBzHc:-OIrnIPmBtQOmu0XBzHd:b2kubvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/da7f90d55df022f89743ca8b057d1a4afcc72923/pkg/sql/create_stats_test.go#L226)):*
> ```Go
> 	b.ResetTimer()
> 	for i := 0; i < b.N; i++ {
> 		sqlRunner.Exec(b, `ANALYZE t`)
> ```

nit: to reduce variance, maybe execute `SELECT count(*) FROM t` query before the benchmark loop? This will ensure that intents are resolved outside of the benchmark loop.
___
*[`pkg/sql/rowexec/sampler.go` line 526 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIrot9wBW5jqghlcrlC:-OIrot9wBW5jqghlcrlD:b-bp6m67) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fdff61e6f8f119d9ee79ca7ba2a887cf356bd8/pkg/sql/rowexec/sampler.go#L526)):*
> ```Go
> 	*buf = (*buf)[:0]
> 	for _, col := range s.spec.Columns {
> 		if b := row[col].EncodedBytes(); b != nil && !containsCollatedString(typs[col]) {
> ```

nit: should we calculate `containsCollatedString` property only once instead of for each row? Perhaps the check is quick enough to not matter - up to you.
___
*[`pkg/sql/rowexec/sampler.go` line 529 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIroM8m4z3p8dgND7kZ:-OIroM8m4z3p8dgND7k_:b-dtsukq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fdff61e6f8f119d9ee79ca7ba2a887cf356bd8/pkg/sql/rowexec/sampler.go#L529)):*
> ```Go
> 			// If the datum is already encoded and does not contain a collated
> 			// string, we can insert the encoded bytes directly into the sketch.
> 			// Even thought the encoded bytes contain the column ID and type,
> ```

nit: `s/Even thought/Even though/`.
___
*[`pkg/sql/rowexec/sampler.go` line 628 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIrntEzCfFI9w4F6CCD:-OIrntEzCfFI9w4F6CCE:b88kbtd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fdff61e6f8f119d9ee79ca7ba2a887cf356bd8/pkg/sql/rowexec/sampler.go#L628)):*
> ```Go
> // legacy implementation from versions prior to 25.2.
> //
> // TODO(mgartner): Remove this once compatibility with 24.3 is no longer needed.
> ```

nit: `s/24.3/25.1/`.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140601)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mw5h and @yuzefovich)

___
*[`pkg/sql/rowexec/sampler.go` line 526 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIrot9wBW5jqghlcrlC:-OIuT1ce1RQ87nXoVemB:b5ibfjw) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fdff61e6f8f119d9ee79ca7ba2a887cf356bd8/pkg/sql/rowexec/sampler.go#L526)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: should we calculate `containsCollatedString` property only once instead of for each row? Perhaps the check is quick enough to not matter - up to you.
</blockquote></details>

Good idea. I'll try that in a follow-up PR. Though I have a feeling the check is very fast compared to inserting into the HLL sketch, so the gains might be small.
___
*[`pkg/sql/rowexec/sampler.go` line 529 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIroM8m4z3p8dgND7kZ:-OIuTNne50yJDr8KURih:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fdff61e6f8f119d9ee79ca7ba2a887cf356bd8/pkg/sql/rowexec/sampler.go#L529)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/Even thought/Even though/`.
</blockquote></details>

Done.
___
*[`pkg/sql/rowexec/sampler.go` line 628 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIrntEzCfFI9w4F6CCD:-OIuTUU55WM3CfqmGNNO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2fdff61e6f8f119d9ee79ca7ba2a887cf356bd8/pkg/sql/rowexec/sampler.go#L628)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: `s/24.3/25.1/`.
</blockquote></details>

Done.
___
*[`pkg/sql/create_stats_test.go` line 226 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIrnIPmBtQOmu0XBzHc:-OIuUm8wE9SkHKaXF0mp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/da7f90d55df022f89743ca8b057d1a4afcc72923/pkg/sql/create_stats_test.go#L226)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: to reduce variance, maybe execute `SELECT count(*) FROM t` query before the benchmark loop? This will ensure that intents are resolved outside of the benchmark loop.
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/distsql_stats` line 3074 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIuXZ7u6KhLycoRZERh:-OIuXZ7u6KhLycoRZERi:b-7mnttb) ([raw file](https://github.com/cockroachdb/cockroach/blob/c471adc5c9a93441ff56b23131e7088ee57dd1e9/pkg/sql/logictest/testdata/logic_test/distsql_stats#L3074)):*
> ```
> SELECT count(*) FROM [SHOW HISTOGRAM $hist_id_int_outer_buckets_full]
> ----
> 200
> ```

This change is worrisome. I need to figure out why it's happening before merging.


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/logictest/testdata/logic_test/distsql_stats` line 3074 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIuXZ7u6KhLycoRZERh:-OIvJ4fWA81v5a9W_xeP:b-rv3tvm) ([raw file](https://github.com/cockroachdb/cockroach/blob/c471adc5c9a93441ff56b23131e7088ee57dd1e9/pkg/sql/logictest/testdata/logic_test/distsql_stats#L3074)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This change is worrisome. I need to figure out why it's happening before merging.
</blockquote></details>

Becuase this changes distinct counts slightly, the outer buckets were not being created any more, causing the test changes. I updated the tests with injected stats to ensure there are outer buckets and make the test less likely to flake.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> This is looking good!

> Reviewed 10 of 10 files at r5, 2 of 2 files at r6, 4 of 7 files at r8, 5 of 7 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-:-OIwcwYlDLkZpKRRRvqm:bov0ylg)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mw5h and @yuzefovich)

___
*[`pkg/sql/create_stats_test.go` line 204 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIw8YN1B_CyVQ4XD1nF:-OIw8YN1B_CyVQ4XD1nG:b-dxw7ia) ([raw file](https://github.com/cockroachdb/cockroach/blob/7f99bcb6c71df41a532744188d97cadd9083ab33/pkg/sql/create_stats_test.go#L204)):*
> ```Go
> 			pad CHAR(60) NOT NULL DEFAULT '',
> 			INDEX (k)
> 		)
> ```

might be worth turning off `kv.split_queue.enabled` or doing something else to ensure there's always only a single range
___
*[`pkg/sql/logictest/testdata/logic_test/distsql_stats` line 3073 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIwcBEo8yzdiSjIJaCf:-OIwcBEo8yzdiSjIJaCg:b-elhhly) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fa99a6d771233fee3d5d23664b369a8a6e2e92c/pkg/sql/logictest/testdata/logic_test/distsql_stats#L3073)):*
> ```
>     "avg_size": 2,
>     "histo_col_type": "int",
>     "histo_buckets": [
> ```

might be good to stick a `"histo_version": 3,` in there, too
___
*[`pkg/sql/logictest/testdata/logic_test/distsql_stats` line 3076 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIwcVfRBzRSNTXWdUsp:-OIwcVfRBzRSNTXWdUsq:b-ew5ji8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fa99a6d771233fee3d5d23664b369a8a6e2e92c/pkg/sql/logictest/testdata/logic_test/distsql_stats#L3076)):*
> ```
>       {"num_eq": 0, "num_range": 0, "distinct_range": 0, "upper_bound": "-9223372036854775808"},
>       {"num_eq": 1, "num_range": 0, "distinct_range": 0, "upper_bound": "0"},
>       {"num_eq": 1, "num_range": 9998, "distinct_range": 0, "upper_bound": "9999"},
> ```

distinct_range should probably be 9998 for this bucket
___
*[`pkg/sql/opt/exec/execbuilder/testdata/partial_stats` line 1259 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIwcEBO9HkLw-eHZrRS:-OIwcEBO9HkLw-eHZrRT:b-1vnx2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fa99a6d771233fee3d5d23664b369a8a6e2e92c/pkg/sql/opt/exec/execbuilder/testdata/partial_stats#L1259)):*
> ```
>     "avg_size": 2,
>     "histo_col_type": "int",
>     "histo_buckets": [
> ```

same comment about histo_version
___
*[`pkg/sql/opt/exec/execbuilder/testdata/partial_stats` line 1262 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIwcdnFC4Xute-mUQ6q:-OIwcdnFC4Xute-mUQ6r:b368cuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fa99a6d771233fee3d5d23664b369a8a6e2e92c/pkg/sql/opt/exec/execbuilder/testdata/partial_stats#L1262)):*
> ```
>       {"num_eq": 0, "num_range": 0, "distinct_range": 0, "upper_bound": "-9223372036854775808"},
>       {"num_eq": 1, "num_range": 0, "distinct_range": 0, "upper_bound": "0"},
>       {"num_eq": 1, "num_range": 9997, "distinct_range": 0, "upper_bound": "9998"},
> ```

distinct_range should be 9997 for this bucket
___
*[`pkg/sql/opt/exec/execbuilder/testdata/partial_stats` line 1323 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIwcsPJ-xp91rOWfk4Z:-OIwcsPJ-xp91rOWfk4_:b-re3qzy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fa99a6d771233fee3d5d23664b369a8a6e2e92c/pkg/sql/opt/exec/execbuilder/testdata/partial_stats#L1323)):*
> ```
> # Verify that distinct counts and null counts are merged correctly.
> statement ok
> ALTER TABLE ka INJECT STATISTICS '[
> ```

same comments for this one
___
*[`pkg/sql/rowexec/sampler.go` line 605 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIwaTFY8j-Yb2JRGZna:-OIwaTFY8j-Yb2JRGZnb:b-waxkjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fa99a6d771233fee3d5d23664b369a8a6e2e92c/pkg/sql/rowexec/sampler.go#L605)):*
> ```Go
> // true with collated string types with a default-equivalent collation.
> func containsCollatedString(t *types.T) bool {
> 	isNonDefaultEquivalentCollatedString := func(t *types.T) bool {
> ```

nit: making `containsCollatedString` recursive would avoid the need for this inner function and would also support arrays of arrays of collated strings or arrays of tuples of collated strings (which I believe are possible in PG)


<!-- Sent from Reviewable.io -->

mgartner commented:
TFTRs!

bors r+

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140601)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale)

___
*[`pkg/sql/rowexec/sampler.go` line 605 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIwaTFY8j-Yb2JRGZna:-OIznz7e7OrJ-8J8HCRe:b-dpmq67) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fa99a6d771233fee3d5d23664b369a8a6e2e92c/pkg/sql/rowexec/sampler.go#L605)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

nit: making `containsCollatedString` recursive would avoid the need for this inner function and would also support arrays of arrays of collated strings or arrays of tuples of collated strings (which I believe are possible in PG)
</blockquote></details>

great catch, done. it's much simpler now
___
*[`pkg/sql/logictest/testdata/logic_test/distsql_stats` line 3073 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIwcBEo8yzdiSjIJaCf:-OIzoLRg9LZDQNn-Pjve:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fa99a6d771233fee3d5d23664b369a8a6e2e92c/pkg/sql/logictest/testdata/logic_test/distsql_stats#L3073)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

might be good to stick a `"histo_version": 3,` in there, too
</blockquote></details>

Done.
___
*[`pkg/sql/logictest/testdata/logic_test/distsql_stats` line 3076 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIwcVfRBzRSNTXWdUsp:-OIzoQij5N14A4kVjvnp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fa99a6d771233fee3d5d23664b369a8a6e2e92c/pkg/sql/logictest/testdata/logic_test/distsql_stats#L3076)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

distinct_range should probably be 9998 for this bucket
</blockquote></details>

Done.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/partial_stats` line 1259 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIwcEBO9HkLw-eHZrRS:-OIzowTw2aPosBEmmRRd:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fa99a6d771233fee3d5d23664b369a8a6e2e92c/pkg/sql/opt/exec/execbuilder/testdata/partial_stats#L1259)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

same comment about histo_version
</blockquote></details>

done.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/partial_stats` line 1262 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIwcdnFC4Xute-mUQ6q:-OIzox3n9cTJEF5spHcG:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fa99a6d771233fee3d5d23664b369a8a6e2e92c/pkg/sql/opt/exec/execbuilder/testdata/partial_stats#L1262)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

distinct_range should be 9997 for this bucket
</blockquote></details>

done.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/partial_stats` line 1323 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/140601#-OIwcsPJ-xp91rOWfk4Z:-OIzoutI6HKmX8sqqOVa:b-7rl0jd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3fa99a6d771233fee3d5d23664b369a8a6e2e92c/pkg/sql/opt/exec/execbuilder/testdata/partial_stats#L1323)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

same comments for this one
</blockquote></details>

done.


<!-- Sent from Reviewable.io -->
